Add scalar wave rhs

nilsdeppe opened the pull request:
## Proposed changes

- Add ability to call specific constructor via `make_array`
- Change to the new `make_array` call in `Tensor`
- Add ScalarWave tags
- Add ScalarWave RHS

I'm planning on adding the flux equations in a separate PR that will be dependent on this PR.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

nilsdeppe committed:
Allow calling constructors through make_array

An overload is added to allow calling non-copy constructor inside
make_array. This is useful for non-copyable types, and when forwarding
initializer_list's in generic code.

nilsdeppe committed:
Use forwarding make_array in Tensor

Also avoid narrowing conversions in calls to Tensor forwarding
constructor.

nilsdeppe committed:
Add ScalarWave Tags and time derivative equations

wthrowe commented on tests/Unit/Utilities/Test_MakeArray.cpp:
```diff
@@ -59,4 +80,24 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
+  // Check make_array with a multi-arg constructor
+  const auto vector_array1 = make_array<3, std::vector<int>>(3_st, 9);
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(vector_array1[0] == (std::vector<int>(3, 9)));
```
Again.

kidder replied:
ditto

wthrowe commented on tests/Unit/Utilities/Test_MakeArray.cpp:
```diff
@@ -59,4 +80,24 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
+  // Check make_array with an initializer list constructor
+  const auto vector_array2 = make_array<3, std::vector<int>>({3, 9, 12, -10});
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(vector_array2[0] == (std::vector<int>{3, 9, 12, -10}));
```
Again.

kidder replied:
ditto

wthrowe commented on tests/Unit/Utilities/Test_MakeArray.cpp:
```diff
@@ -59,4 +80,24 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
+      make_array<3, MyNonCopyable<std::vector<double>>>(
+          std::vector<double>{2.3, 8.6, 9.7});
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(noncopyable_vectors[0].get() == (std::vector<double>{2.3, 8.6, 9.7}));
```
Shouldn't this depend on the loop index somehow?

kidder replied:
shouldn't this check all the elements of the array? i.e. `[0]` -> `[i]`

wthrowe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -0,0 +1,60 @@
+ * \frac{\partial^2 u}{\partial t^2} - \frac{\partial^2 u}{\partial x^2}=0
+ * \f]
+ * See Hesthaven and Warburton, Nodal Discontinuous Galerkin Methods, page
+ * 159, or wikipedia.
```
`Wikipedia` (proper noun)

kidder commented on src/Utilities/MakeArray.hpp:
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
-      {(static_cast<void>(indexes), value)..., std::forward<T>(value)}};
-}
-// clang-format on
+template <bool SizeZero>
```
please update the comment; there is no `indexes` or `value` or `size`

kidder commented on src/Utilities/MakeArray.hpp:
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
+ * \ingroup UtilitiesGroup
+ * \brief Create a `std::array<T, size>{{T(args...), T(args...), ...}}`
+ * \tparam Size the size of the array
+ * \tparam T the type of the element is the array
```
is -> in

kidder commented on src/Utilities/MakeArray.hpp:
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
-  return std::array<std::decay_t<T>, 0>{{}};
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Create a `std::array<T, size>{{T(args...), T(args...), ...}}`
```
size -> Size

wthrowe commented on src/Evolution/Systems/ScalarWave/System.hpp:
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Variables.hpp"
```
These first two includes are unused.

kidder commented on src/Utilities/MakeArray.hpp:
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
- * \brief Helper class to initialize a std::array.
- *
- * \tparam size the length of the array
+ * \brief Create a `std::array<std::decay_t<T>, size>{{t, t, ...}}`
```
size -> Size

kidder commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -0,0 +1,60 @@
+ * \f[
+ * \frac{\partial^2 u}{\partial t^2} - \frac{\partial^2 u}{\partial x^2}=0
+ * \f]
+ * See Hesthaven and Warburton, Nodal Discontinuous Galerkin Methods, page
```
I think this comment should describe the first-order form of the scalar wave equation

kidder commented on src/Evolution/Systems/ScalarWave/Tags.hpp:
```diff
@@ -0,0 +1,28 @@
+
+#pragma once
+
+#include "DataStructures/DataBoxTag.hpp"
```
includes (or forward declarations) for size_t and DataVector

kidder commented on src/Evolution/Systems/ScalarWave/System.hpp:
```diff
@@ -0,0 +1,29 @@
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/ScalarWave/Equations.hpp"
```
first two not yet used

kidder commented on src/Evolution/Systems/ScalarWave/System.hpp:
```diff
@@ -0,0 +1,29 @@
+/// Defines class ScalarWaveSystem.
+
+#pragma once
+
```
include for size_t

kidder commented on src/Evolution/Systems/ScalarWave/System.hpp:
```diff
@@ -0,0 +1,29 @@
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace ScalarWave {
```
does this namespace have a doxygen comment anywhere?

kidder commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Utilities/Gsl.hpp"
```
forward declaration sufficient?

nilsdeppe committed:
Add EvolutionSystemsGroup doxygen group

nilsdeppe committed:
Remove ParalelComponentsGroup doxygen group

nilsdeppe committed:
Add GeneralizedHarmonic system to EvolutionSystemsGroup

nilsdeppe commented:
While addressing the comment about documenting the ScalarWave namespace I added a three commits. One adds a Doxygen group for evolution systems, another removes the currently unused and non-alphabetical ParallelComponentsGroup, and the last one adds GH to the evolution systems group.

kidder commented on src/Utilities/MakeArray.hpp:
```diff
@@ -19,14 +19,15 @@
-/// place, rather than undergoing another copy.
-/// Order of evaluation is well-defined for aggregate initialization, so there
-/// is no risk of copy-after-move
+// We handle the zero size and case separately below because both for zero size
```
no and

kidder commented on docs/GroupDefs.hpp:
```diff
@@ -90,6 +90,11 @@
 
+/*!
+ * \defgroup EvolutionSystemsGroup Evolution Systems
+ * \brief Contains namespace of all the available evolution systems.
```
the namespaces

kidder reviewed (changes requested):
as far as I am concerned you can squash when making the latest changes
